### PR TITLE
[sailfish-secrets] Add C API for Sailfish Secrets. Contributes to JB#36797

### PR DESCRIPTION
**Pull request by chriadam**
This commit adds a C API which can be used by non-C++ applications to
perform some simple operations with the Sailfish Secrets API.

### CODE REVIEW

**Comment by chriadam** [comment]
The Secrets C API offered is purely synchronous, limited to a subset of the C++ functionality, and currently there are some important caveats (e.g. only latin1/ascii strings are supported...) but should suffice for the cases where it's needed (that is, system daemons written in C with very limited requirements).

I still need to implement the Crypto C API.

**Comment by Denis-Semakin** [comment]
As I can see the low-level dbus API (libdbus) is used in this pull-request. In my opinion I would advise to use Glib2 DBus API (https://developer.gnome.org/gio/stable/GDBusConnection.html) in project which is written in C (many examples can be found in Google). It allows to write less source code and it provides a small high-level abstraction layer.

Also, I've take a look at PR a little bit and want to notice that a good practice in C is that one should check the returned result of every single function which was called. Especially it related for functions working with dynamic memory(malloc, free, etc).
The code 
-----------------8<-----------------
while (curr) {
    free(curr->some_pointer);
}
-----------------8<----------------- is not guarantee that curr->some_pointer is valid, it may be NULLed :).
Check the code please. 
I would pay attention oh how it may be done using GLib2, one may use their g_new* functions (https://developer.gnome.org/glib/2.28/glib-Memory-Allocation.html#glib-Memory-Allocation.description)
so you can be sure that memory is allocated and provided by the system and can be successfully freed without any crashes.

**Comment by monich** [comment]
> Also, I've take a look at PR a little bit and want to notice that a good practice in C is that one should check the returned result of every single function which was called. Especially it related for functions working with dynamic memory(malloc, free, etc).

Checking the result of a memory allocation function makes no sense on a system configured for overcommit (which is pretty much every Linux system these days). Basically with overcommit on, memory allocation never fails unless you are running out of virtual address space. The actual allocation happens when the page is first accessed and if you are running of memory, your program just dies with SIGBUS or something.

Also, the g_new macros/functions that you mentioned, abort the program in an unlikely event that memory allocation fails.

Finally, in order for memory allocation checks to make any sense and be theoretically of any use, those have to be performed and handled everywhere in the whole program. On top of that, overcommit has to be turned off. That's really really really hard to imagine. 99.99999% of the code written to recover from memory allocation failures in user space programs is dead code which never gets executed.

The bottom line: memory allocation checks make sense in drivers, very little sense in user space programs and practically no sense in Linux user space environment.

**Comment by monich** [comment]
> As I can see the low-level dbus API (libdbus) is used in this pull-request. In my opinion I would advise to use Glib2 DBus API (https://developer.gnome.org/gio/stable/GDBusConnection.html) in project which is written in C (many examples can be found in Google). It allows to write less source code and it provides a small high-level abstraction layer.

I would use gio too but in the end it's a matter of taste. As long as the interface doesn't enforce dependency on gio or libdbus (which seems to be the case), it's an implementation detail and as such it's up to the implementer. There are many ways to do the same thing, gio and libdbus are not clearly better or worse than each other, they are just different. Although, like I said, I would vote for gio too.

**Comment by Denis-Semakin** [comment]
I just wanted to say that checking of successfully or failed memory allocation is a good practice for C programming, may be it will never called, may be, but if it will occurred - we'll have a code to handle it.
I would stay a paranoiac in this case.

Moreover there might be a multiple heavy user applications running on device which cat use all available memory (do we have any cgroups limits for this?). Device (user-space application) might be hacked and fork-bombed and checking will lead to avoid the situation of crashing a daemon.

I've done a several tests on SoC systems when a memory has been exhausted and the "malloc" really returned NULL. Do you want to say this is impossible situation? :)

File: lib/Secrets/secrets_c.h [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    size_t dataSize;
+};
+
+struct Sailfish_Secrets_Secret_Identifier*
+Sailfish_Secrets_Secret_Identifier_new(
+        const char *name,
+        const char *collectionName);
+
+void Sailfish_Secrets_Secret_Identifier_delete(
+        struct Sailfish_Secrets_Secret_Identifier *ident);
```
Comment (monich) [inline lib/Secrets/secrets_c.h]: I would strongly suggest to make objects and any potentially shareable data structures ref-countable (i.e. use new/ref/unref style API). That makes life so much easier and costs so little on the implementation side that I can't think of any reason not to do it.

File: lib/Secrets/secrets_c.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        enum Sailfish_Secrets_SecretManager_UserInteractionMode uiMode,
+        const char *interactionServiceAddress,
+        struct Sailfish_Secrets_Result **out_result);
+
+int Sailfish_Secrets_SecretManager_getSecret(
+        struct Sailfish_Secrets_Secret_Identifier *ident,
+        enum Sailfish_Secrets_SecretManager_UserInteractionMode uiMode,
+        const char *interactionServiceAddress,
+        struct Sailfish_Secrets_Result **out_result,
+        struct Sailfish_Secrets_Secret **out_secret);
```
Comment (monich) [inline lib/Secrets/secrets_c.h]: Most operation (certainly get/set) can block, right? Then cancelable asynchronous version of each such call is needed too. I would even say required.

File: lib/Secrets/secrets_c.c [hunk 3]
```diff
@@ -0,0 +1,1065 @@
...
+        dbus_error_free(&err);
+        return 0;
+    }
+
+    /* send message and get a handle for a reply */
+    dbus_connection_send_with_reply(
+                *sessionBus,
+                msg,
+                &pending,
+                -1);
```
Comment (monich) [inline lib/Secrets/secrets_c.c]: Blocking call detected! And there are more all over the place. That's no good.

**Comment by monich** [comment]
> I've done a several tests on SoC systems when a memory has been exhausted and the "malloc" really returned NULL. Do you want to say this is impossible situation? :)

I didn't say that it never makes sense, I said that it makes practically no sense in Linux user space environment because it's going to be dead code that never runs in real life. And if the code never runs, it's typically full of bugs unless it gets extensively unit-tested. Extensively unit-testing dead code makes even less sense.

I mentioned drivers as an example of when it it does make sense. Yes, one can add various RTOS and weird SoC programming environments to that list too.

**Comment by vityafx** [comment]
@monich What about unrecoverable errors such as `malloc` failure and some invalid state of a program, for example, having a file opened and not fully written? For example, we could check `malloc` for nullptr and if it is, then properly perform the cleaning up operations and exit with a code which we can understand. Also, I have had one malloc failure recently on the phone and my check performed very well - cleaned everything up and exited the app.

Though, if the OOM-killer kills in 100% cases, it is really no need to check for the return value of the `malloc`, in my opinion.

**Comment by Denis-Semakin** [comment]
Avoiding malloc's results checking will mean that we are consciously shooting yourself in a foot.

**Comment by monich** [comment]
Don't get me wrong, I agree that checking the result of malloc is a good habit. It's just that sometimes it makes no sense and in those cases it's OK to consciously skip that.

File: lib/Crypto/crypto_c.cpp [hunk 4]
```diff
@@ -0,0 +1,1523 @@
...
+    if (daemon_connection.p2pBus == NULL) {
+        return 0;
+    }
+
+    return 1;
+}
+
+void Sailfish_Crypto__disconnectFromServer()
+{
+    dbus_connection_close(daemon_connection.p2pBus);
```
[the remark this answers is not in the document]
Comment (monich) [reply lib/Crypto/crypto_c.cpp]: If GObjects were used, then private D-Bus connection could be closed automatically when all references are released (using the weak reference mechanism) which would make this function totally unnecessary. Also, what happens if the other side of the connection dies?

File: lib/Crypto/crypto_c.h [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+enum Sailfish_Crypto_Key_Operation {
+    Sailfish_Crypto_Key_OperationUnknown    = 0,
+    Sailfish_Crypto_Key_Sign                = 1,
+    Sailfish_Crypto_Key_Verify              = 2,
+    Sailfish_Crypto_Key_Encrypt             = 4,
+    Sailfish_Crypto_Key_Decrypt             = 8
+};
+
+struct Sailfish_Crypto_Key {
+    struct Sailfish_Crypto_Key_Identifier *identifier;
```
[the remark this answers is not in the document]
Comment (monich) [reply lib/Crypto/crypto_c.h]: Padding only makes sense if this structure is supposed to be allocated by the client of the library. If it's only allocated by the library itself then padding is unnecessary - old clients just won't know that the structure is bigger than they thought and won't access the new fields. Direct access to the public fields is easier, more readable and more efficient than a function call, I'm personally fine with that (and do that myself quite often). Obviously, new fields have to be added to the end of the structure.

File: lib/Secrets/secrets_c.c [hunk 6]
```diff
@@ -0,0 +1,1078 @@
...
+                DBUS_TYPE_INT32,
+                arg);
+    dbus_message_iter_close_container(
+                in_args,
+                enum_struct);
+}
+
+void Sailfish_Secrets__appendSecretIdentifierArg(
+        DBusMessageIter *args,
+        struct Sailfish_Secrets_Secret_Identifier *ident)
```
Comment (dcaliste) [inline lib/Secrets/secrets_c.c]: One may want to factorise code here between secret and crypto. The identifiers are the same struct of two char* fields, they maybe shared and DBUS code handling these can be shared too.

File: lib/Crypto/crypto_c.cpp [hunk 7]
```diff
@@ -0,0 +1,1523 @@
...
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_STRING) {
+            fprintf(stderr, "Fourth result struct arg not a string!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &errorMessage);
+
+        *result = Sailfish_Crypto_Result_new(
+                    (Sailfish_Crypto_Result_Code)code,
+                    errorCode, storageErrorCode, errorMessage);
```
[the remark this answers is not in the document]
Comment (dcaliste) [reply lib/Crypto/crypto_c.cpp]: Ok, forget this comment, checking https://dbus.freedesktop.org/doc/api/html/group__DBusMessage.html#ga41c23a05e552d0574d0444d4693d18ab there is no memory leak.

To avoid thinking this when reading the code, would the compiler allow to declare errorMessage as a const char* instead of a char* ?

File: lib/Crypto/crypto_c.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (dcaliste) [reply lib/Crypto/crypto_c.h]: You're right for the padding. GObject is using it because it is inserting parent struct in child struct and they need to keep the offset right for child when adding new attributes in parent. It's not the case here. So no padding!

**Comment by dcaliste** [comment]
I'm wondering also (I didn't read the C++ API sorry), would it be useful to add a cryptoManager_storedKeys() function to get the list of available keys from the storage? I mean, currently we can sign/encrypt/verify/decrypt by providing a key as argument, and the key can be retrieved from the storage with the cryptoManager_storedKey() method when knowing the identifier of the key. But if we are designing (in C) something to present the available list of keys for signing purposes and then choose from the list. Currently we cannot. GPG is providing this, see https://www.gnupg.org/documentation/manuals/gpgme/Listing-Keys.html#Listing-Keys for instance.

In the case you agree, what would be the API for such method, specifically, how to handle a list of keys? Like in GPG with a next pointer at the end of the key structure, or in a GList way, dedicating a new struct for this?

File: lib/Crypto/crypto_c.h [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+struct Sailfish_Crypto_Key_CustomParameter*
+Sailfish_Crypto_Key_CustomParameter_new(
+        const unsigned char *parameter,
+        size_t parameterSize);
+
+void Sailfish_Crypto_Key_CustomParameter_delete(
+        struct Sailfish_Crypto_Key_CustomParameter *param);
+
+struct Sailfish_Crypto_Key*
+Sailfish_Crypto_Key_new(
```
Comment (dcaliste) [inline lib/Crypto/crypto_c.h]: Would we want the C user to create new keys by him or the only way to obtain keys are using the cryptoManager functions? Should this new() function be moved out of the public header?

File: lib/Secrets/secrets_c.c (hunk 3 above)
Comment (chriadam) [reply lib/Secrets/secrets_c.c]: Yes, the C API is currently purely synchronous.  Yes, we should add asynchronous API.  I just created this quick API as a proof of concept and something we can iterate on.

File: lib/Crypto/crypto_c.cpp [hunk 9]
```diff
@@ -0,0 +1,1523 @@
...
+    DBusMessageIter args;
+    DBusError err;
+    DBusPendingCall* pending;
+    char *param = NULL;
+
+    /* initialise the error */
+    dbus_error_init(&err);
+
+    /* connect to the session bus */
+    *sessionBus = dbus_bus_get(DBUS_BUS_SESSION, &err);
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: The public functions do all internally check whether a connection to the daemon has been established, and if not, attempts to connect.  But you are correct that that depends on the availability of the session bus in order to get the discoverable p2p bus address, I will double check that we handle errors in the discovery stage properly.  It certainly shouldn't segfault, but instead should return -1 and print an error string to stderr.

File: lib/Crypto/crypto_c.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: Thanks, I'll add a check for p2pBus state here.

Regarding GObjects, I'd prefer to keep the dependencies super minimal if possible.  Most clients should use the C++ API, IMO.  But we can look into using glib and gdbus if that would be best for the most likely clients of this C API (i.e. ofono and connman).  Patches would be most welcome, as my glib programming experience is quite limited.

File: lib/Crypto/crypto_c.cpp [hunk 10]
```diff
@@ -0,0 +1,1523 @@
...
+        dbus_message_iter_next(&res_struct);
+
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_STRING) {
+            fprintf(stderr, "Fourth result struct arg not a string!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &errorMessage);
+
+        *result = Sailfish_Crypto_Result_new(
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: Regarding the error message, I wanted this API to mirror the C++ API as closely as possible, and in that one the Result is provided to the client in all cases.  Explicitly, it's not an error, but a result, and can be Succeeded/Pending/Failed.

Regarding the check for result: I agree, this sort of defensive check is required.  I will add it.

File: lib/Crypto/crypto_c.h [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+struct Sailfish_Crypto_Result {
+    enum Sailfish_Crypto_Result_Code code;
+    int errorCode;
+    int storageErrorCode;
+    char *errorMessage;
+};
+
+struct Sailfish_Crypto_Result*
+Sailfish_Crypto_Result_new(
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.h]: I provided it purely for symmetry, but you are correct that clients should not need it.  I will move it to private implementation.

File: lib/Crypto/crypto_c.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.h]: My only real opinion about this is: we shouldn't commit to supporting this C API with BC/SC guarantees until we have real clients using the C++ API (including things which currently aren't fleshed out, but are very important, like certificates), because currently I see this API as being entirely WIP / subject to change.

Basically, this API exists as a prototype only and I expect it to evolve over the coming weeks as we actually start to implement real use-cases with the APIs.

Aside from that, my experience with pure C APIs is obviously quite limited compared to my experience with C++ APIs so I will no doubt have made a lot of mistakes / many improvements can be made / patches welcome ;-)

File: lib/Crypto/crypto_c.h [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    Sailfish_Crypto_Key_Encrypt             = 4,
+    Sailfish_Crypto_Key_Decrypt             = 8
+};
+
+struct Sailfish_Crypto_Key {
+    struct Sailfish_Crypto_Key_Identifier *identifier;
+
+    enum Sailfish_Crypto_Key_Origin origin;
+    enum Sailfish_Crypto_Key_Algorithm algorithm;
+    int blockModes;
```
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.h]: It's a flag, so it can be the result of ORing several BlockMode enum values.

File: lib/Secrets/secrets_c.c (hunk 6 above)
Comment (chriadam) [reply lib/Secrets/secrets_c.c]: This was an explicit decision, to keep the two libraries entirely separate (that is, clients which link to libsailfishsecrets do not have to link to libsailfishcrypto and vice versa).  I guess we could use some private implementation code sharing inclusion tricks to share implementation but for such a small struct, I don't see the point, really.

File: lib/Crypto/crypto_c.h (hunk 8 above)
Comment (chriadam) [reply lib/Crypto/crypto_c.h]: Client applications can create new keys themselves, or they can create new template keys and ask a crypto plugin to generate a key based on the template, or they can create a reference key (which contains just a valid identifier).  Some might argue that the first use-case needn't be supported, but certainly the last two do need to be supported.  So I think we need this method public.

**Comment by Denis-Semakin** [comment]
A general question: why do you use such C-unlike coding style? IMHO.
Wouldn't be better and friendly to use kernel coding style for all C-project?

**Comment by chriadam** [comment]
Regarding malloc checks, I do agree we should add those.

However, the case that you pointed out regarding "free(curr->next)" isn't a problem, because if curr->next is NULL, the man page states that free(NULL) is a no-op.

I'm open to using gdbus instead of libdbus, however I don't have much experience using it myself.  It may be that using gdbus will greatly simplify the code if we need asynchronous APIs, and if so, porting this code to that may make sense.  Otherwise, I'd prefer to not add more complicated dependencies.  But let's see.

**Comment by chriadam** [comment]
Regarding the style - partially it's because I wanted the API to be symmetric with the C++ API, and partially it's no doubt due to the fact that I have far more experience with C++ and Qt APIs than with C APIs.  Patches or suggestions for improvement are welcome, of course!

File: lib/Crypto/crypto_c.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: I've added invalid parameter check preamble to the public functions now.

File: lib/Crypto/crypto_c.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (dcaliste) [reply lib/Crypto/crypto_c.h]: Ok, I see your point. But, if I take GLib as example, they are using enum for ORed flags and the type that is used is always the enum type itself, see for instance: https://developer.gnome.org/gobject/stable/GBinding.html#g-object-bind-property

File: lib/Crypto/crypto_c.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: Done.

File: lib/Crypto/crypto_c.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (monich) [reply lib/Crypto/crypto_c.h]: gdb deals nicely with OR'ed enums (showing symbolic names rather than numeric value)

**Comment by monich** [comment]
> A general question: why do you use such C-unlike coding style? IMHO.
> Wouldn't be better and friendly to use kernel coding style for all C-project?

Nobody would argue that it's better to use his favourite style for all projects :) Let's leave it up to the author and maintainer. The style looks strange to me too but at least it's consistent.

File: lib/Secrets/secrets_c.h (hunk 1 above)
Comment (chriadam) [reply lib/Secrets/secrets_c.h]: Ok, I'll add the reference count and ref/unref methods, and remove the delete methods

File: lib/Crypto/crypto_c.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (chriadam) [reply lib/Crypto/crypto_c.cpp]: The invalid parameter check preamble also checks these cases.

**Comment by monich** [comment]
BTW, C libraries shouldn't depend on C++ runtime, let alone Qt.

**Comment by chriadam** [comment]
You're right.  We should extract these C APIs to a separate small library then.  There is only one thing stopping that happening immediately: the Key serialisation/deserialisation methods.  We may need to write those in terms of primitives instead of relying on QDataStream serialisation, I think, so that we can appropriately reimplement them within the C API impl.

In any event, I think that can wait until we have an actual need for these C APIs.

**Comment by izh1979** [comment]
>> A general question: why do you use such C-unlike coding style? IMHO.
>> Wouldn't be better and friendly to use kernel coding style for all C-project?

> Nobody would argue that it's better to use his favourite style for all projects :) Let's leave it up to the  author and maintainer. The style looks strange to me too but at least it's consistent.

IMHO, the problem is that Sailfish Code Style doesn't cover internal C projects. It says how to write C++ apps only. If we define a unified style for C sources, it would be great. But personally I vote for the kernel style.

File: lib/Secrets/secrets_c.h (hunk 1 above)
Comment (chriadam) [reply lib/Secrets/secrets_c.h]: Should I just add a (publicly visible) int refcount; member to each struct?  One alternative would be to add a void* member (which I allocate manually, but that guarantees another heap allocation).  A final alternative would be to only provide access/setter functions, and just expose the struct as an opaque handle, but that's suboptimal for a few reasons IMO...

Comment (monich) [reply lib/Secrets/secrets_c.h]: You missed another alternative! [Here](https://git.merproject.org/mer-core/libdbusaccess/blob/master/src/dbusaccess_peer.c#L51) is an example. Basically, the real (internal) structure is bigger than the public part, the internal part containing refcount or whatever else you don't want to expose. There are examples of similar technique in glib itself (e.g. GArray). That's what I like the most (except when I need a GObject, then the approach is slightly different).

**Comment by chriadam** [comment]
So, current plan for this one: I will rewrite these with gio / gdbus to provide asynchronous API, and separate them out of the C++ libs and into a libsailfishsecretscryptoc library which avoids pulling in Qt etc dependencies.  I will (while doing that) provide better checking for malloc return values etc.

**Comment by vityafx** [comment]
@chriadam sounds good to me.

**Comment by chriadam** [comment]
First draft of the gdbus-based Secrets API: https://github.com/sailfishos/sailfish-secrets/pull/12 - Crypto API is still TODO.  I haven't yet done the malloc checking, either, but it should be kernel style (brace style and indentation, perhaps not variable naming however - that is TODO also).  I'm on vacation for a few weeks, back mid-January.

**Comment by chriadam** [comment]
Superseded by the work by inz
